docs/ideviceenterrecovery.1
.TH "ideviceenterrecovery" 1
.SH NAME
ideviceenterrecovery \- Make a device enter recovery mode.
.SH SYNOPSIS
.B ideviceenterrecovery
[OPTIONS] UDID

.SH DESCRIPTION

Makes a device with the supplied UDID enter recovery mode immediately.

.SH OPTIONS
.TP
.B \-d, \-\-debug
enable communication debugging.
.TP 
.B \-h, \-\-help
prints usage information.
.TP
.B \-v, \-\-version
prints version information.

.SH AUTHORS
Martin Szulecki

Man page written to conform with Debian by Julien Lavergne.

.SH ON THE WEB
https://libimobiledevice.org

https://github.com/libimobiledevice/libimobiledevice


NEWS
Version 1.3.0
~~~~~~~~~~~~~

* Development release "Let's get the party started."
* Changes:
  - Fix Python 3 support
  - Add more lockdown error codes
  - Add new lockdownd_pair_with_options() function
  - Fix GnuTLS support with iOS 10
  - Make sure sockets only listen locally due to security reasons
  - Plug various memory leaks
  - Fix SSL version negotiation for newer versions of OpenSSL
  - Optimize lockdown pair record handling
  - Return proper error code when a lockdown pair record is missing
  - Fix building with MingGW
  - Store application information in Info.plist using idevicebackup2
  - Fix application backup handling to allow the device to restore applications
    that were installed using idevicebackup2
  - Make idevicebackup2 reboot after restore the default to allow the device to
    migrate data correctly and thus improve the restored device data state
  - Improve console frontend information output in idevicebackup2
  - Extend ideviceprovision tool to allow retrieving and removing all
    provisioning profiles
  - Fix parsing large provisioning profile using ideviceprovision
  - Fix receiving large property lists in property list service
  - Propagate lower level errors to callers instead of returning
    IDEVICE_E_UNKNOWN_ERROR
  - API: Add IDEVICE_DEVICE_PAIRED event type
  - Detect screenshot format to support png, tiff and dat formats using
    idevicescreenshot tool
  - API: Add mobileactivation service implementation
  - Wait for passcode entry if required using idevicesyslog
  - Add HDMI option to diagnostics command for idevicediagnostics
  - Fix IORegistry command for iOS 11+ devices in idevicediagnostics
  - Remove 40-digit character limit for UDID in tools to support newer devices
  - Fix broken validate command in idevicepair with iOS 11+
  - Fix OpenSSL version checks for configure target when using LibreSSL
  - Migrate latest improved common code from libusbmuxd
  - Convert README file to markdown format
  - Fix idevicecrashreport tool to work with iOS 13+
  - Fix various errors in SSL communication logic
  - API: Add preboard service implementation
  - Output hint to user to enter passcode when changing password using
    idevicebackup2
  - Cython: Fix and improve debugserver and diagnostics service bindings
  - API: Add WiFi device support via new idevice_new_with_options() function
  - API: Add idevice_get_device_list_extended() to also list network devices
  - API: Add lockdown_strerror() helper to get error representation as string
  - Add network device support to idevicesyslog and ideviceinfo tools
  - Make debug output consistently output to stderr
  - Add new idevicesetlocation tool (requires mounted developer image)
  - Add option to exit if device disconnects in idevicesyslog
  - API: Add syslog_relay_start_capture_raw() for raw syslog capture
  - Add color output and process filter support to idevicesyslog
  - API: Add companion_proxy service implementation
  - Bump dependency to libusbmuxd 2.0.2
  - Bump dependency to libplist 2.2.0
  - Improve error handling and reporting in library and tools
  - Fix various memory leaks in library and tools
  - Add "--network" and "--version" options to all tools
  - Fix socket_connect_addr() not connecting to network devices using IPv6
    in some cases.
  - Improve IPv6 "scope id" detection to fix connecting to network devices with
    link-local adresses.
  - Update man pages
  - Fix various inconsistent declarations in public headers
  - Allow OpenSSL >= 1.1.0 to use older/disallowed TLS versions fixing issues
    where pairing records were getting removed repeatingly
  - Fixed memory leaks
  - Cython: Rewrite version detection logic in configure.ac
  - Rename "--enable-debug-code" configure option to "--enable-debug"
  - Improve README.md with project description, installation, contributing and
    usage sections
  - Rename library and all related files by adding an API version resulting
    in "libimobiledevice-1.0"
  - Bump soname version
* API is UNSTABLE

Version 1.2.0
~~~~~~~~~~~~~

* Stable release "It took you so long baby!"
* Changes:
  - Require autoconf 2.64 or later
  - Remove dev tools, will return either as proper tools or website examples
  - Refactor installation proxy service implementation and normalize code
  - API: Added instproxy_lookup() to efficiently lookup app information
  - API: Added instproxy_check_capabilities_match() to check device capabilities
  - API: Added various instproxy command and status plist getters
  - API: Make debugserver_client_set_ack_mode() public
  - Fix handling of clients reconnecting in idevicedebugserverproxy which
    previously didn't work properly
  - Flush stdout for every line in idevicesyslog
  - Fix shutdown of idevicedebugserverproxy tool which could hang
  - Notify user when erroneously using idevicebackup with iOS 4 or later
  - Enable build of idevicecrashreport on WIN32
  - Fix thread handle leaks on WIN32 adding thread_new and thread_free
  - cython: Add receive/receive_timeout methods for iDeviceConnection to
    receive raw data from a connection
  - cython: Add new FILE_RELAY_E_PERMISSION_DENIED(-6) error
  - API: Refactor lockdown service internal error checking and add a bunch of
    new native errors
  - Convert int16_t macro error types into enum within common module, too
  - Add new "idevicenotificationproxy" tool to post or observe notifications
  - Fix overlong blocking in np_client_free()
  - Improve maintainability and Requires of pkg-config file
  - API: Add new LOCKDOWN_E_SERVICE_LIMIT error to detect service limit states
  - API: Remove const argv requirement for debugserver_command_new
  - cython: Add get_path_for_bundle_identifier() method to
    InstallationProxyClient
  - cython: Add DebugServerClient class to communicate with debugserver
  - Comply to strict function prototypes by using (void) instead of just ()
  - Fix notification proxy shutdown process which was incorrectly implemented

tools/ideviceenterrecovery.c
/*
 * ideviceenterrecovery.c
 * Simple utility to make a device in normal mode enter recovery mode.
 *
 * Copyright (c) 2009 Martin Szulecki All Rights Reserved.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#define TOOL_NAME "ideviceenterrecovery"

#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <stdlib.h>
#ifndef WIN32
#include <signal.h>
#endif

#include <libimobiledevice/libimobiledevice.h>
#include <libimobiledevice/lockdown.h>

static void print_usage(int argc, char **argv)
{
	char *name = NULL;

	name = strrchr(argv[0], '/');
	printf("Usage: %s [OPTIONS] UDID\n", (name ? name + 1: argv[0]));
	printf("\n");
	printf("Makes a device with the supplied UDID enter recovery mode immediately.\n");
	printf("\n");
	printf("OPTIONS:\n");
	printf("  -d, --debug\t\tenable communication debugging\n");
	printf("  -h, --help\t\tprints usage information\n");
	printf("  -v, --version\t\tprints version information\n");
	printf("\n");
	printf("Homepage:    <" PACKAGE_URL ">\n");
	printf("Bug Reports: <" PACKAGE_BUGREPORT ">\n");
}

int main(int argc, char *argv[])
{
	lockdownd_client_t client = NULL;
	lockdownd_error_t ldret = LOCKDOWN_E_UNKNOWN_ERROR;
	idevice_t device = NULL;
	idevice_error_t ret = IDEVICE_E_UNKNOWN_ERROR;
	int i;
	const char* udid = NULL;

#ifndef WIN32
	signal(SIGPIPE, SIG_IGN);
#endif
	/* parse cmdline args */
	for (i = 1; i < argc; i++) {
		if (!strcmp(argv[i], "-d") || !strcmp(argv[i], "--debug")) {
			idevice_set_debug_level(1);
			continue;
		}
		else if (!strcmp(argv[i], "-h") || !strcmp(argv[i], "--help")) {
			print_usage(argc, argv);
			return 0;
		}
		else if (!strcmp(argv[i], "-v") || !strcmp(argv[i], "--version")) {
			printf("%s %s\n", TOOL_NAME, PACKAGE_VERSION);
			return 0;
		}
	}

	i--;
	if (argc < 2 || !argv[i] || !*argv[i]) {
		print_usage(argc, argv);
		return 0;
	}
	udid = argv[i];

	ret = idevice_new(&device, udid);
	if (ret != IDEVICE_E_SUCCESS) {
		printf("No device found with udid %s.\n", udid);
		return -1;
	}

	if (LOCKDOWN_E_SUCCESS != (ldret = lockdownd_client_new(device, &client, TOOL_NAME))) {
		printf("ERROR: Could not connect to lockdownd, error code %d\n", ldret);
		idevice_free(device);
		return -1;
	}

	/* run query and output information */
	printf("Telling device with udid %s to enter recovery mode.\n", udid);
	if(lockdownd_enter_recovery(client) != LOCKDOWN_E_SUCCESS)
	{
		printf("Failed to enter recovery mode.\n");
	}
	printf("Device is successfully switching to recovery mode.\n");

	lockdownd_client_free(client);
	idevice_free(device);

	return 0;
}
